### PR TITLE
Feature/sling 7768: String Interpolation for ETC Map

### PR DESCRIPTION
**Pull request by schaefa**
Fix for ticket: https://issues.apache.org/jira/browse/SLING-7768 providing a way to make /etc/map configurable for different environments.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -63,6 +63,32 @@
                     </execution>
                 </executions>
             </plugin>
+<!--
```
Comment (cziegeler) [inline pom.xml]: Changes in the build should not be part of this PR - unless they are strictly required

File: src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverFactoryActivator.java [hunk 2]
```diff
@@ -107,6 +109,19 @@
     @Reference
     EventAdmin eventAdmin;
 
+    /** Event admin. */
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverFactoryActivator.java]: Comment above seems to be wrong :) Please make the reference greedy and remove the commented out lines below

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
+/*
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java]: It's ok to have a copy here for getting the functionality in, but we should rather directly use the class from Apache Felix by embedding just that class (and using the maven shade plugin to move it to a different package). But we can do this in a follow up commit

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/MapEntries.java [hunk 4]
```diff
@@ -140,8 +142,10 @@
 
     private boolean updateBloomFilterFile = false;
 
+    private StringInterpolationProvider stringInterpolationProvider;
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/mapping/MapEntries.java]: This can be marked as final

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+)
+public @interface StringInterpolationProviderConfiguration {
+
+    @AttributeDefinition(
+        name = "Placeholder Values",
+        description = "A list of key / value pairs separated by a equal (=) sign. " +
+            "The key is not permitted to contain a '=' sign and the first occurrence of '=' " +
+            "separates the key from the value. If no '=' is found the entire key / value pair " +
+            "is dropped.")
+    String[] placeHolderKeyValuePairs() default {"phv.default.host.name=localhost"};
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java]: Do we really need such a default value?

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.metatype.annotations.Designate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Designate(ocd = StringInterpolationProviderConfiguration.class)
+@Component(name = "org.apache.sling.resourceresolver.impl.mapping.StringInterpolationProviderImpl")
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: name attribute is not needed

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    @Override
+    public String substitute(String text) {
+        logger.trace("Substitute: '{}'", text);
+        Object result = Interpolator.replace(text, (type, name, dir) -> {
+            String v = null;
+            if (TYPE_ENV.equals(type)) {
+                v = getVariableFromEnvironment(name);
+            } else if (TYPE_PROP.equals(type)) {
+                v = getVariableFromProperty(name);
+            } else {
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: we should define a type for this, like "config:" and not let all types fallback to this. The fallback prevents adding new types later on (if required)

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    private static final String TYPE_ENV = "env";
+
+    private static final String TYPE_PROP = "prop";
+
+    private static final String DIRECTIVE_DEFAULT = "default";
+
+    /** Logger. */
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    private Map<String, String> placeholderEntries = new HashMap<>();
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: Can be marked as final

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    private BundleContext context;
+
+    // ---------- SCR Integration ---------------------------------------------
+
+    /**
+     * Activates this component (called by SCR before)
+     */
+    @Activate
```
Comment (cziegeler) [inline src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: I suggest to change this to a constructor

File: src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java [hunk 10]
```diff
@@ -428,15 +429,15 @@ private Resource buildResource(String fullpath, Iterable<Resource> children, Res
...
-    /**
-     * extract the name from a path.
-     * @param fullpath
-     * @return
-     */
-    private String getName(String fullpath) {
-        int n = fullpath.lastIndexOf("/");
-        return fullpath.substring(n+1);
-    }
+//    /**
```
Comment (cziegeler) [inline src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java]: Please remove the commented code

File: src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java [hunk 11]
```diff
@@ -399,7 +400,7 @@ private Resource buildResource(String fullpath, Iterable<Resource> children) {
     @SuppressWarnings("unchecked")
     private Resource buildResource(String fullpath, Iterable<Resource> children, ResourceResolver resourceResolver, ResourceProvider<?> provider, String ... properties) {
         Resource resource = Mockito.mock(Resource.class);
-        Mockito.when(resource.getName()).thenReturn(getName(fullpath));
+        Mockito.when(resource.getName()).thenReturn(getResourceName(fullpath));
```
Comment (cziegeler) [inline src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java]: ResourceUtil.getName() ?

File: src/test/java/org/apache/sling/resourceresolver/impl/mapping/EtcMappingMapEntriesTest.java [hunk 12]
```diff
@@ -16,12 +16,40 @@
  */
 package org.apache.sling.resourceresolver.impl.mapping;
```
Comment (cziegeler) [inline src/test/java/org/apache/sling/resourceresolver/impl/mapping/EtcMappingMapEntriesTest.java]: There are a lot of changes in this test and I can't see whether these are just improvements or tests need to be adapted?

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java (hunk 3 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java]: The class is not exported from the Bundle (AFAI can tell) and so I could not use it.

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java (hunk 5 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java]: That was here because I wanted to illustrate it. I am going to remove it.

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java (hunk 7 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: I did not like the fallback either but I wanted to avoid forcing the user to write 'config:' for each value. That said etc-mapping is not easy by any chance and so I guess we add that.

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java (hunk 8 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: Thanks for point this out

File: src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java (hunk 11 above)
Comment (schaefa) [reply src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java]: I merge in some changes from master to make it easier to merge and so this is not a change of mine.

File: src/test/java/org/apache/sling/resourceresolver/impl/mapping/EtcMappingMapEntriesTest.java (hunk 12 above)
Comment (schaefa) [reply src/test/java/org/apache/sling/resourceresolver/impl/mapping/EtcMappingMapEntriesTest.java]: As far as I can tell I did not change anything here beside that I might have removed commented out code.

File: src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java (hunk 10 above)
Comment (schaefa) [reply src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java]: Not my change but will do

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java (hunk 9 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: You want to make the activate method part of the constructor? Can OCD being part of the constructor?

File: pom.xml (hunk 1 above)
Comment (schaefa) [reply pom.xml]: Done

File: src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverFactoryActivator.java (hunk 2 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverFactoryActivator.java]: Done

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java (hunk 3 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java]: I reverted everything back to the code from Felix with the exception of added logging statements and constants

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/MapEntries.java (hunk 4 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/MapEntries.java]: Done

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java (hunk 5 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java]: Done

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java (hunk 6 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: Done

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java (hunk 7 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: Done. This required the tests to be adjusted as now $[config: is required.

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java (hunk 8 above)
Comment (schaefa) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderImpl.java]: Opted to replace the Map in activate (modified)

File: src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java (hunk 10 above)
Comment (schaefa) [reply src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java]: Done

File: src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java (hunk 11 above)
Comment (schaefa) [reply src/test/java/org/apache/sling/resourceresolver/impl/MockedResourceResolverImplTest.java]: Kept it as this was not my change

File: src/test/java/org/apache/sling/resourceresolver/impl/mapping/EtcMappingMapEntriesTest.java (hunk 12 above)
Comment (schaefa) [reply src/test/java/org/apache/sling/resourceresolver/impl/mapping/EtcMappingMapEntriesTest.java]: There are some additional tests cases to deal with types and defaults but the rest did change little.

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java (hunk 3 above)
Comment (cziegeler) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/Interpolator.java]: We can embed it, exactly because it is not exported - but there is no strict need for it, other than to avoid code duplication - and as mentioned we can do this later on

File: src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java (hunk 5 above)
Comment (cziegeler) [reply src/main/java/org/apache/sling/resourceresolver/impl/mapping/StringInterpolationProviderConfiguration.java]: I still see it :)
